### PR TITLE
Don't recurse down closures for duplicate-label checking

### LINKED ISSUE #25343: Incorrect warning about label shadowing itself
The following code:
```
fn main() {
    || {
        'label: loop {
        }
    };
}

```

Results in this warning:
```
main.rs:3:17: 4:10 warning: label name `'label` shadows a label name that is already in scope
main.rs:3         'label: loop {
main.rs:4         }
main.rs:3:17: 4:10 note: shadowed label `'label` declared here
main.rs:3         'label: loop {
main.rs:4         }
```

Obviously a label can't shadow itself, so there shouldn't be a warning. The issue doesn't occur outside of a closure.

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #25343

To be honest I'm not sure if this is the right fix (I haven't yet fully understood the code),
but it seems to work. I'll look closer at the code when I have some time, in the meantime if this
is the right fix it would be nice to get verification from someone who does understand the code :smile:

r? @pnkfelix

### CODE REVIEW

**Comment by Manishearth** [comment]
cc @Ms2ger

File: src/test/run-pass/issue-25343.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    || {
+        'label: loop {
+        }
+    };
+}
```
Comment (tamird) [inline src/test/run-pass/issue-25343.rs]: this will fail tidy

Comment (tamird) [inline src/test/run-pass/issue-25343.rs]: oh, apparently not?

**Pull request merged**